lib/datasets/custom_voc_structure_db.py

```diff
@@ -0,0 +1,148 @@
+        imdb.__init__(self, name, class_set)
+        # Parameters
+        self.config = {'use_salt': True,  # generic parameter: use uuid to salt comp_id
+                       'cleanup': True,  # generic parameter: remove temp files created for evaluations
+                       'min_size': 2,     # selective search: filter small boxes
+                       'top_k': 2000,  # region proposals: top K proposal to consider
+                       'use_diff': False,  # VOC-specific: exclude samples labeled as difficult
+                       }
+        # name, paths
+        if devkit_path is None:
+            self._data_path = os.path.dirname(image_set_txt_path)
+            self._image_set_txt = os.path.basename(image_set_txt_path)
+        else:
+            self._data_path = devkit_path
+            self._image_set_txt = image_set_txt_path
+
+        self._image_index = self._load_image_set_index()
+        # TODO add support for .png?
+        self._image_exts = ['.JPEG', '.JPG', '.jpg']
+        self._roidb_handler = self.gt_roidb
+        self.competition_mode(False)
+
+    # NOTE: modified from pascal_voc.py to handle multiple image
+    # suffixes, not just '.jpg'
+    def image_path_from_index(self, index):
+        """
+        Construct an image path from the image's "index" identifier.
+        """
+        image_path = ''
+        for _image_ext in self._image_exts:
+            image_path = os.path.join(self._data_path, 'JPEGImages',
+                                      index + _image_ext)
+            if os.path.exists(image_path):
+                break
+        assert os.path.exists(image_path), \
+            'Path does not exist: {}'.format(image_path)
+        return image_path
+
+    # NOTE: modified from pascal_voc.py
+    def _load_image_set_index(self):
+        """
+        Load the indexes listed in this dataset's image set file.
+        """
+        if os.path.isfile(self._image_set_txt):
+            # The path leads directly to a file, so read it in.
+            image_set_file = self._image_set_txt
+        else:
+            # image_set_txt does not reference a file, so it must be relative to
+            # data_path.
+            image_set_file = os.path.join(self._data_path, self._image_set_txt)
+        assert os.path.exists(image_set_file), \
+            'Path does not exist: {}'.format(image_set_file)
+        with open(image_set_file) as f:
+            image_index = [x.strip() for x in f.readlines()]
+        return image_index
+
+    # NOTE: modified from pascal_voc.py, so imitates VOC-style 1-indexing
+    def _load_pascal_annotation(self, index):
+        """
+        Load image and bounding boxes info from XML file in the PASCAL VOC
+        format.
+        """
+        filename = os.path.join(self._data_path, 'Annotations', index + '.xml')
+        return self._load_pascal_annotation_xml(filename)
+
+    def _load_pascal_annotation_xml(self, xmlpath):
```

Comment:
It's not great for maintainability that this function is pretty much an exact duplicate of the one from `pascal_voc`. I guess it's not a bit deal if we aren't going to be doing any more work on this repo, but I think it would have been cleaner to modify `datasets.pascal_voc` such that its `load_pascal_annotation` function could be reused here.